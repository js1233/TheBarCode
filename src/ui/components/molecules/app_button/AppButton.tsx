import { FONT_SIZE, SPACE } from "config";
import Colors from "config/Colors";
import { usePreferredTheme } from "hooks";
import React, { useEffect, useState } from "react";
import {
  ActivityIndicator,
  ImageStyle,
  StyleProp,
  StyleSheet,
  TextStyle,
  TouchableOpacity,
  TouchableOpacityProps,
  View,
  ViewStyle
} from "react-native";
import {
  AppLabel,
  TEXT_TYPE
} from "ui/components/atoms/app_label/AppLabel";
import { SvgProp } from "utils/Util";

export interface AppButtonProps extends TouchableOpacityProps {
  onPress?: () => void;
  text?: string;
  buttonStyle?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  textDisableStyle?: StyleProp<TextStyle>;
  textContainerStyle?: StyleProp<ViewStyle>;
  shouldShowProgressBar?: boolean;
  loaderSize?: number;
  loaderColor?: string;
  leftIcon?: SvgProp | undefined;
  rightIcon?: SvgProp;
  buttonType?: BUTTON_TYPES;
  isDisable?: boolean;
  buttonDisableStyle?: StyleProp<ViewStyle>;
  iconStyle?: StyleProp<ImageStyle>;
  shouldShowError?: boolean;
  textType?: TEXT_TYPE;
  shouldNotOptimize?: boolean;
  shouldAlignTextWithLeftIconWithFullWidth?: boolean;
}

export enum BUTTON_TYPES {
  NORMAL = "normal",
  BORDER = "border",
  DASH = "dashed"
}

export const AppButton = React.memo<AppButtonProps>(
  ({
    text,
    onPress,
    iconStyle,
    buttonStyle,
    textStyle,
    textDisableStyle,
    shouldShowProgressBar = false,
    loaderSize = 15,
    loaderColor,
    leftIcon,
    rightIcon,
    buttonType = BUTTON_TYPES.NORMAL,
    isDisable = false,
    buttonDisableStyle,
    shouldShowError = false,
    textType = TEXT_TYPE.NORMAL,
    shouldAlignTextWithLeftIconWithFullWidth = false,
    textContainerStyle
  }) => {
    const theme = usePreferredTheme();
    const [customDisabled, setIsDisabled] = useState<boolean>(isDisable);

    useEffect(() => {
      let timeout: any;
      if (isDisable) {
        setIsDisabled(isDisable);
      } else {
        //when we need to close screen after button click, changing isDisable to false
        //trigger a second click due to whic hapi is called again to prevent this {isDisbaled:false}
        //is calling after a little delay, to properly close the parent screen in the meantime
        timeout = setTimeout(() => {
          setIsDisabled(isDisable);
        }, 300);
      }

      () => clearTimeout(timeout);
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isDisable]);

    const getButtonStyle = () => {
      if (buttonType === BUTTON_TYPES.DASH) {
        return [
          style.buttonWithDash,
          { borderColor: theme.themedColors.primaryColor }
        ];
      } else if (buttonType === BUTTON_TYPES.BORDER) {
        return [
          style.buttonWithBorder,
          {
            borderColor: shouldShowError
              ? Colors.colors.red
              : theme.themedColors.primaryColor
          }
        ];
      }
    };

    const getShadowStyle = () => {
      if (!customDisabled) {
        return [
          style.buttonWithShadow,
          { shadowColor: theme.themedColors.primaryBackground }
        ];
      }
    };

    return (
      <TouchableOpacity
        onPress={shouldShowProgressBar ? undefined : onPress}
        disabled={customDisabled}
        style={[
          style.button,
          getShadowStyle(),
          getButtonStyle(),
          {
            backgroundColor: theme.themedColors.primaryColor
          },
          customDisabled
            ? buttonDisableStyle ?? [
                buttonStyle,
                { backgroundColor: theme.themedColors.interface[400] }
              ]
            : buttonStyle
        ]}>
        <View testID="button-container" style={style.viewContainer}>
          {leftIcon && !shouldShowProgressBar && (
            <View
              style={[
                iconStyle,
                style.leftIconContainer,
                !shouldAlignTextWithLeftIconWithFullWidth
                  ? style.leftIconContainerPosition
                  : undefined
              ]}>
              {leftIcon?.(theme.themedColors.primaryColor, 20, 20)}
            </View>
          )}
          <View
            style={[
              style.textWithLoader,
              !shouldAlignTextWithLeftIconWithFullWidth
                ? style.textWithLoaderFlex
                : undefined,
              textContainerStyle
            ]}>
            {!shouldShowProgressBar && (
              <AppLabel
                style={[
                  style.text,
                  customDisabled ? textDisableStyle : textStyle
                ]}
                text={text}
                textType={textType}
                allowFontScaling={false}
              />
            )}
            {shouldShowProgressBar && (
              <ActivityIndicator
                testID="loader"
                style={[style.loader]}
                size={loaderSize}
                color={
                  loaderColor ?? theme.themedColors.secondaryBackground
                }
              />
            )}
          </View>
          {rightIcon && !shouldShowProgressBar && (
            <View style={style.rightIconContainer}>
              {rightIcon?.(theme.themedColors.primaryColor, 20, 20)}
            </View>
          )}
        </View>
      </TouchableOpacity>
    );
  }
);

const style = StyleSheet.create({
  buttonWithShadow: {
    elevation: 1,
    shadowOffset: { width: 0, height: 3 },
    shadowRadius: 3,
    shadowOpacity: 0.15
  },
  text: {
    fontSize: FONT_SIZE.base,
    overflow: "hidden",
    color: Colors.colors.white
  },
  loader: {
    marginLeft: 10
  },
  textWithLoader: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center"
  },
  textWithLoaderFlex: {
    flex: 1
  },
  leftIconContainer: {
    alignItems: "flex-start",
    marginLeft: SPACE.sm
  },
  leftIconContainerPosition: {
    position: "absolute",
    left: SPACE.sm
  },
  leftIcon: {
    marginLeft: 10,
    width: 20,
    height: 20
  },
  rightIconContainer: {
    position: "absolute",
    right: SPACE.lg,
    justifyContent: "flex-end",
    alignItems: "flex-end"
  },
  rightIcon: {
    marginRight: 10,
    width: 20,
    height: 20
  },
  buttonWithBorder: {
    borderWidth: 1,
    borderStyle: "solid"
  },
  buttonWithDash: {
    borderWidth: 1,
    borderStyle: "dashed"
  },
  button: {
    alignSelf: "center",
    flexDirection: "row",
    width: "100%",
    minHeight: 44,
    maxHeight: 55,
    justifyContent: "space-between",
    alignItems: "center",
    borderRadius: 25
  },
  viewContainer: {
    flexDirection: "row",
    alignItems: "center",
    padding: SPACE.sm
  }
});
