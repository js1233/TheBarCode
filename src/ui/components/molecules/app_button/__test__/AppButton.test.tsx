import { fireEvent, render } from "@testing-library/react-native";
import React from "react";
import "react-native";
import { ReactTestInstance } from "react-test-renderer";
import { AppButton } from "ui/components/molecules/app_button/AppButton";
import LeftIcon from "assets/images/left.svg";
import RightIcon from "assets/images/right.svg";

test("snapshot testing", () => {
  const rendered = render(<AppButton text="Submit" />).toJSON();
  expect(rendered).toMatchSnapshot();
});

test("should properly perform click event", () => {
  const onPress = jest.fn();
  const { getByText } = render(
    <AppButton text="Submit" onPress={onPress} />
  );
  const button = getByText("Submit");
  fireEvent.press(button);
  expect(onPress).toBeCalledTimes(1);
});

it("renders activity indicator correctly", () => {
  let showPb = true;
  const { queryByTestId, update } = render(
    <AppButton text="Submit" shouldShowProgressBar={showPb} />
  );
  let loader = queryByTestId("loader");
  expect(loader).not.toBeNull();

  showPb = false;
  update(<AppButton text="Submit" shouldShowProgressBar={showPb} />);
  loader = queryByTestId("loader");
  expect(loader).toBeNull();
});

it("renders left icon", () => {
  const leftSvgIcon = () => {
    return <LeftIcon testID="left-icon" width={20} height={20} />;
  };
  const { queryByTestId } = render(
    <AppButton text="Submit" leftIcon={leftSvgIcon} />
  );
  const leftIconId = queryByTestId("left-icon");
  const rightIconId = queryByTestId("right-icon");
  expect(rightIconId).toBeNull();
  expect(leftIconId).not.toBeNull();
});

it("works fine even when all hell broke loose", async () => {
  const rightSvgIcon = () => {
    return <RightIcon testID="right-icon" width={20} height={20} />;
  };
  // given
  let rightIcon: ReactTestInstance | null = null,
    leftIcon: ReactTestInstance | null = null,
    loader: ReactTestInstance | null = null,
    text: ReactTestInstance | null = null;

  function queryComponents() {
    rightIcon = queryByTestId("right-icon");
    leftIcon = queryByTestId("left-icon");
    loader = queryByTestId("loader");
    text = queryByTestId("title");
  }

  function changePbStatus(_onPress: any, _update: any, _showPb: boolean) {
    showPb = _showPb;
    _update(
      <AppButton
        text="Submit"
        onPress={_onPress}
        shouldShowProgressBar={showPb}
        rightIcon={rightSvgIcon}
      />
    );
  }

  let showPb = false;
  const onPress = jest.fn((_onPress, _update) => {
    changePbStatus(_onPress, _update, true);
  });

  // when
  const { queryByTestId, getByText, update } = render(
    <AppButton
      text="Submit"
      onPress={() => {
        onPress(onPress, update);
      }}
      shouldShowProgressBar={showPb}
      rightIcon={rightSvgIcon}
    />
  );

  queryComponents();
  let button = getByText("Submit");

  // then
  expect(leftIcon).toBeNull();
  expect(rightIcon).not.toBeNull();
  expect(loader).toBeNull();
  expect(onPress).toBeCalledTimes(0);

  // when
  fireEvent.press(button);

  // when
  queryComponents();

  // then
  expect(leftIcon).toBeNull();
  expect(rightIcon).toBeNull();
  expect(text).toBeNull();
  expect(loader).not.toBeNull();
  expect(onPress).toBeCalledTimes(1);

  // when
  changePbStatus(onPress, update, false);
  queryComponents();

  // then
  expect(leftIcon).toBeNull();
  expect(rightIcon).not.toBeNull();
  expect(loader).toBeNull();
  expect(onPress).toBeCalledTimes(1);
});

test("onPress should not called in case of disable button", () => {
  const onPress = jest.fn();
  const { getByText } = render(
    <AppButton text="Submit" onPress={onPress} isDisable={true} />
  );
  const button = getByText("Submit");
  fireEvent.press(button);
  expect(onPress).toBeCalledTimes(0);
});
