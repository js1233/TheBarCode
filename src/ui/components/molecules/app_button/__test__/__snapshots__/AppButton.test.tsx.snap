// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "#0F766E",
      "borderRadius": 25,
      "elevation": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "maxHeight": 55,
      "minHeight": 44,
      "opacity": 1,
      "shadowColor": "#FFFFFF",
      "shadowOffset": Object {
        "height": 3,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 3,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "padding": 8,
      }
    }
    testID="button-container"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          },
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <Text
        allowFontScaling={false}
        numberOfLines={1}
        style={
          Array [
            Object {
              "fontFamily": "ProximaNova-Regular",
            },
            Object {
              "color": "#171717",
              "fontSize": 27,
            },
            Object {
              "lineHeight": 42,
            },
            Array [
              Object {
                "color": "#FFFFFF",
                "fontSize": 28,
                "overflow": "hidden",
              },
              undefined,
            ],
          ]
        }
      >
        Submit
      </Text>
    </View>
  </View>
</View>
`;
